articles/contributing/api-design-principles.md

```diff
@@ -0,0 +1,426 @@
+      applied inputs occur first (i.e.: so that partial application
+      acts similarly to currying).
+
+    - For example, an operation ApplyRotation that takes a
+        floating-point number and a qubit as inputs may often be
+        partially applied with the floating-point input first for
+        use with operations that expect an input of type
+        `Qubit => Unit`. Thus, a signature of
+        `operation ApplyRotation(angle : Double, target : Qubit) : Unit is Adj + Ctl`
+        would work most consistently with partial application.
+
+    - Typically, this guidance means placing all classical data
+        before all qubits in input tuples, but use good judgment and
+        examine how your API is called in practice.
+
+## User-Defined Types
+
+- **DO** use user-defined types to help make APIs more expressive and
+    convenient to use.
+
+  - **DO** introduce new user-defined types to provide helpful
+      shorthand for long and/or complicated types. For example:
+
+    - In cases where an operation type with three qubit array
+        inputs is commonly taken as an input or returned as an
+        output, providing a UDT such as
+        `newtype TimeDependentBlockEncoding = ((Qubit[], Qubit[], Qubit[]) => Unit is Adj + Ctl)`
+        can help provide a useful shorthand.
+
+    - **DO** introduce new user-defined types to indicate that a given
+        base type should only be used in a very particular sense. For
+        example:
+
+      - An operation that should be interpreted specifically as an
+          operation that encodes classical data into a quantum
+          register may be appropriate to label with a user-defined
+          type `newtype InputEncoder = (Apply : (Qubit[] => Unit))`.
+
+    - **DO** introduce new user-defined types with named items that
+        allow for future extensibility (e.g.: a results structure that
+        may contain additional named items in the future). For example:
+
+      - When an operation `TrainModel` exposes a large number of
+          configuration options, exposing these options as a new
+          `TrainingOptions` UDT and providing a new function
+          `DefaultTrainingOptions : Unit -> TrainingOptions` allows
+          users to override specific named items in TrainingOptions
+          UDT values while still allowing library developers to add
+          new UDT items as appropriate.
+
+    - **DO** declare named items for new user-defined types in
+        preference to requiring users to know the correct tuple
+        deconstruction. For example:
+
+      - When representing a complex number in its polar
+          decomposition, prefer
+          `newtype ComplexPolar = (Magnitude: Double, Argument: Double)` to
+          `newtype ComplexPolar = (Double, Double)`.
+
+- **DON'T** use user-defined types in ways that increase cognitive
+    load or that require the user to learn additional concepts and
+    nomenclature.
```

Comment:
Well, this seems to contradict the *introduce new user-defined types to provide helpful shorthand for long and/or complicated types.*.
In your example you define the `TimeDependentBlockEncoding`, but that requires the user to learn and understand what a time dependent block encoding is.